Add lazy loading using `loading="lazy"`

ericcornelissen opened the pull request:
Addresses #37 by replacing the inline `<svg>`s with `<img>`s with the SVG Base-64 encoded on the `src` and `loading="lazy"`. This makes it so that the images are not loaded until they're in view. If `loading="lazy"` is not supported by the browser all icons are loaded immediately (same behaviour as current `master`).

For more details on the trade-offs of various approached we can take, see [this comment](https://github.com/simple-icons/simple-icons-website/pull/55#issuecomment-826394479).

### Preview

https://user-images.githubusercontent.com/3742559/116010501-9f324180-a61f-11eb-950f-44e41893919c.mp4

ericcornelissen committed:
Use `loading="lazy"` on image elements to lazy load icons

And the copy functionality has been updated to `fetch` the raw SVG. From
local testing (Firefox only) it seems this request is always served from
the cache, rather actually sending a second request to get the SVG.

service-paradis commented on public/index.pug:
```diff
@@ -125,9 +125,7 @@ html
-                svg(id=`${icon.slug}-svg` role="img" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg")
-                  title #{icon.title} icon
-                  path(d=icon.path)
+                img(src=`./icons/${icon.slug}.svg` loading="lazy")
```
Great!
Another downside that I did not think before: I think the `fill` property will not work anymore...
I think, the icon wont be filled in white in dark mode. Cannot try it right now though.

ericcornelissen replied:
Good point, you're correct. Thanks for pointing it out :+1: 

I think this can be solved using plain HTML, let me investigate :thinking:

service-paradis replied:
The possible alternative would be to use `img` with data url (svg encoded in base64). Something like this:
```
<img src="data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHZpZXdCb3g9IjAgMCAyNCAyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48dGl0bGU+U2ltcGxlIEljb25zIGljb248L3RpdGxlPjxwYXRoIGQ9Ik0xMiAwQzguNjg4IDAgNiAyLjY4OCA2IDZzMi42ODggNiA2IDZjMS4wNjYgMCAyLjExNzEuMjg2MiAzLjAzNzEuODI2MiAyLjg1OCAxLjY3OCAzLjgxNjcgNS4zNTM5IDIuMTM4NyA4LjIxMjloMS42Nzk3YTcuNDkyNSA3LjQ5MjUgMCAwIDAgLjY0MjUtMy4wMjkzYy4wMDMtNC4xNDQtMy4zNTYtNy41MDQ4LTcuNDk4LTcuNTA5OC0yLjQ4NCAwLTQuNS0yLjAxNi00LjUtNC41UzkuNTE2IDEuNSAxMiAxLjVzNC41IDIuMDE2IDQuNSA0LjVIMThjMC0zLjMxMi0yLjY4OC02LTYtNnptMCAzYy0xLjU5Ny4wNC0yLjg3OTkgMS4zMjU5LTIuOTIxOSAyLjkyMTlDOS4wMzUxIDcuNTc5OSAxMC4zNDMgOC45NTcgMTIgOWMxLjU5Ny0uMDQgMi44Nzk5LTEuMzI1OSAyLjkyMTktMi45MjE5QzE0Ljk2NDkgNC40MjAxIDEzLjY1NiAzLjA0MyAxMiAzem0wIDEuNWMuODI4IDAgMS41LjY3MiAxLjUgMS41di4wMDJjMCAuODI4LS42NzIgMS41LTEuNSAxLjUtLjgzIDAtMS41LS42NzMtMS41LTEuNTAyIDAtLjgzLjY3LTEuNSAxLjUtMS41ek03LjUgMTV2MS41SDl2Nkg0LjVWMjRoMTV2LTEuNUgxNXYtNmgxLjVWMTVoLTl6bTMgMS41aDN2NmgtM3YtNnptLTYgMS40NzA3VjE4YTcuNDE4IDcuNDE4IDAgMCAwIC42NDQ1IDMuMDM5aDEuNjgzNkM2LjMwMDEgMjAuMTQ3IDYgMTkuMTEgNiAxOHYtLjAyOTNINC41WiIvPjwvc3ZnPg=="  loading="lazy" />
```
```
<img src="data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHZpZXdCb3g9IjAgMCAyNCAyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48dGl0bGU+U2ltcGxlIEljb25zIGljb248L3RpdGxlPjxwYXRoIGZpbGw9IiNmZmYiIGQ9Ik0xMiAwQzguNjg4IDAgNiAyLjY4OCA2IDZzMi42ODggNiA2IDZjMS4wNjYgMCAyLjExNzEuMjg2MiAzLjAzNzEuODI2MiAyLjg1OCAxLjY3OCAzLjgxNjcgNS4zNTM5IDIuMTM4NyA4LjIxMjloMS42Nzk3YTcuNDkyNSA3LjQ5MjUgMCAwIDAgLjY0MjUtMy4wMjkzYy4wMDMtNC4xNDQtMy4zNTYtNy41MDQ4LTcuNDk4LTcuNTA5OC0yLjQ4NCAwLTQuNS0yLjAxNi00LjUtNC41UzkuNTE2IDEuNSAxMiAxLjVzNC41IDIuMDE2IDQuNSA0LjVIMThjMC0zLjMxMi0yLjY4OC02LTYtNnptMCAzYy0xLjU5Ny4wNC0yLjg3OTkgMS4zMjU5LTIuOTIxOSAyLjkyMTlDOS4wMzUxIDcuNTc5OSAxMC4zNDMgOC45NTcgMTIgOWMxLjU5Ny0uMDQgMi44Nzk5LTEuMzI1OSAyLjkyMTktMi45MjE5QzE0Ljk2NDkgNC40MjAxIDEzLjY1NiAzLjA0MyAxMiAzem0wIDEuNWMuODI4IDAgMS41LjY3MiAxLjUgMS41di4wMDJjMCAuODI4LS42NzIgMS41LTEuNSAxLjUtLjgzIDAtMS41LS42NzMtMS41LTEuNTAyIDAtLjgzLjY3LTEuNSAxLjUtMS41ek03LjUgMTV2MS41SDl2Nkg0LjVWMjRoMTV2LTEuNUgxNXYtNmgxLjVWMTVoLTl6bTMgMS41aDN2NmgtM3YtNnptLTYgMS40NzA3VjE4YTcuNDE4IDcuNDE4IDAgMCAwIC42NDQ1IDMuMDM5aDEuNjgzNkM2LjMwMDEgMjAuMTQ3IDYgMTkuMTEgNiAxOHYtLjAyOTNINC41WiIvPjwvc3ZnPg=="  loading="lazy" />
```
That would give
![image](https://user-images.githubusercontent.com/12817388/112840283-24502800-906d-11eb-99ec-c2ec44ef0d20.png)

ericcornelissen replied:
I "solved" it, though it required CSS. My initial idea was to use an approach similar to what you can find in [this article](https://flaviocopes.com/dark-mode-change-img-source/), but that doesn't work with the _forced_ dark mode and light mode :disappointed: 

If your browser supports lazy loading and JS is enabled, this solution works great. It only loads the dark mode icons once you go into dark mode - and again the cache is used whenever possible.

The downside is that if neither of these are true, the browser will end up downloading **all** SVGs **twice** when you load the website which is really, really bad...

ericcornelissen replied:
> The possible alternative would be to use `img` with data url (svg encoded in base64). Something like this:
> 
> ```
> <img src="data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHZpZXdCb3g9IjAgMCAyNCAyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48dGl0bGU+U2ltcGxlIEljb25zIGljb248L3RpdGxlPjxwYXRoIGQ9Ik0xMiAwQzguNjg4IDAgNiAyLjY4OCA2IDZzMi42ODggNiA2IDZjMS4wNjYgMCAyLjExNzEuMjg2MiAzLjAzNzEuODI2MiAyLjg1OCAxLjY3OCAzLjgxNjcgNS4zNTM5IDIuMTM4NyA4LjIxMjloMS42Nzk3YTcuNDkyNSA3LjQ5MjUgMCAwIDAgLjY0MjUtMy4wMjkzYy4wMDMtNC4xNDQtMy4zNTYtNy41MDQ4LTcuNDk4LTcuNTA5OC0yLjQ4NCAwLTQuNS0yLjAxNi00LjUtNC41UzkuNTE2IDEuNSAxMiAxLjVzNC41IDIuMDE2IDQuNSA0LjVIMThjMC0zLjMxMi0yLjY4OC02LTYtNnptMCAzYy0xLjU5Ny4wNC0yLjg3OTkgMS4zMjU5LTIuOTIxOSAyLjkyMTlDOS4wMzUxIDcuNTc5OSAxMC4zNDMgOC45NTcgMTIgOWMxLjU5Ny0uMDQgMi44Nzk5LTEuMzI1OSAyLjkyMTktMi45MjE5QzE0Ljk2NDkgNC40MjAxIDEzLjY1NiAzLjA0MyAxMiAzem0wIDEuNWMuODI4IDAgMS41LjY3MiAxLjUgMS41di4wMDJjMCAuODI4LS42NzIgMS41LTEuNSAxLjUtLjgzIDAtMS41LS42NzMtMS41LTEuNTAyIDAtLjgzLjY3LTEuNSAxLjUtMS41ek03LjUgMTV2MS41SDl2Nkg0LjVWMjRoMTV2LTEuNUgxNXYtNmgxLjVWMTVoLTl6bTMgMS41aDN2NmgtM3YtNnptLTYgMS40NzA3VjE4YTcuNDE4IDcuNDE4IDAgMCAwIC42NDQ1IDMuMDM5aDEuNjgzNkM2LjMwMDEgMjAuMTQ3IDYgMTkuMTEgNiAxOHYtLjAyOTNINC41WiIvPjwvc3ZnPg=="  loading="lazy" />
> ```
> 
> ```
> <img src="data:image/svg+xml;base64,PHN2ZyByb2xlPSJpbWciIHZpZXdCb3g9IjAgMCAyNCAyNCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48dGl0bGU+U2ltcGxlIEljb25zIGljb248L3RpdGxlPjxwYXRoIGZpbGw9IiNmZmYiIGQ9Ik0xMiAwQzguNjg4IDAgNiAyLjY4OCA2IDZzMi42ODggNiA2IDZjMS4wNjYgMCAyLjExNzEuMjg2MiAzLjAzNzEuODI2MiAyLjg1OCAxLjY3OCAzLjgxNjcgNS4zNTM5IDIuMTM4NyA4LjIxMjloMS42Nzk3YTcuNDkyNSA3LjQ5MjUgMCAwIDAgLjY0MjUtMy4wMjkzYy4wMDMtNC4xNDQtMy4zNTYtNy41MDQ4LTcuNDk4LTcuNTA5OC0yLjQ4NCAwLTQuNS0yLjAxNi00LjUtNC41UzkuNTE2IDEuNSAxMiAxLjVzNC41IDIuMDE2IDQuNSA0LjVIMThjMC0zLjMxMi0yLjY4OC02LTYtNnptMCAzYy0xLjU5Ny4wNC0yLjg3OTkgMS4zMjU5LTIuOTIxOSAyLjkyMTlDOS4wMzUxIDcuNTc5OSAxMC4zNDMgOC45NTcgMTIgOWMxLjU5Ny0uMDQgMi44Nzk5LTEuMzI1OSAyLjkyMTktMi45MjE5QzE0Ljk2NDkgNC40MjAxIDEzLjY1NiAzLjA0MyAxMiAzem0wIDEuNWMuODI4IDAgMS41LjY3MiAxLjUgMS41di4wMDJjMCAuODI4LS42NzIgMS41LTEuNSAxLjUtLjgzIDAtMS41LS42NzMtMS41LTEuNTAyIDAtLjgzLjY3LTEuNSAxLjUtMS41ek03LjUgMTV2MS41SDl2Nkg0LjVWMjRoMTV2LTEuNUgxNXYtNmgxLjVWMTVoLTl6bTMgMS41aDN2NmgtM3YtNnptLTYgMS40NzA3VjE4YTcuNDE4IDcuNDE4IDAgMCAwIC42NDQ1IDMuMDM5aDEuNjgzNkM2LjMwMDEgMjAuMTQ3IDYgMTkuMTEgNiAxOHYtLjAyOTNINC41WiIvPjwvc3ZnPg=="  loading="lazy" />
> ```
> 
> That would give
> ![image](https://user-images.githubusercontent.com/12817388/112840283-24502800-906d-11eb-99ec-c2ec44ef0d20.png)

Sorry, missed this comment. That defeats the entire goal of what we're trying to do here, which is lazy loading the SVGs. If we provide the SVG source directly in the `src` attribute of the SVG we might as well embed the SVG itself - which is what we already have on `master`.

service-paradis replied:
> The downside is that if neither of these are true, the browser will end up downloading all SVGs twice when you load the website which is really, really bad...

Yep, that is why I was trying to find another alternative.

> That defeats the entire goal of what we're trying to do here, which is lazy loading the SVGs.

I think SVGs will still be « lazy rendered », so it might not defeats entirely the goal of this PR. But I understand that the HTML will still be large. Would surely need some test since the behavior of lazy loading inline images is not well documented.

service-paradis replied:
I might have a compromise.
If you only add the default black svg and apply the following style on dark mode, it might do it:
```css
    filter: invert(1);
```
The support is not bad (https://caniuse.com/css-filters)

ericcornelissen replied:
It has better support than `loading="lazy"` itself so worth a try, but indeed it's less than ideal.

ericcornelissen replied:
> I think SVGs will still be « lazy rendered », so it might not defeats entirely the goal of this PR. But I understand that the HTML will still be large. Would surely need some test since the behavior of lazy loading inline images is not well documented.

It does indeed seem to be the case that the icons are still "lazy rendered" and it does indeed seem to still improve performance. I implemented this approach in ba07d08dbba9aa3dc138c7681e55d09af808b003.



> I might have a compromise.
> If you only add the default black svg and apply the following style on dark mode, it might do it:
> ```css
>     filter: invert(1);
> ```
> The support is not bad (https://caniuse.com/css-filters)

As this also works with external SVGs, I gave that a try as well. This fixes the problem of having to load twice as many SVGs in order to support dark mode. I implemented this approach in f8a62fc6f0871b22eba09a3b30eec3cc9677a224


---


Comparing the two approaches above, my findings are as follows:

- Both approaches are faster than the current website. A rough conservative estimate of mine is that they're both around 40% faster...
- But with external SVGs the initial load is slightly faster (if lazy loading is supported). For me the difference seems to be a little less than a second when compared to the "lazy render" approach.
- However, using external SVGs is significantly worse when lazy loading is not supported.
- Using the "lazy render" approach if lazy loading is not supported, the performance is about equal with the current website (as expected).
- If you start scrolling, external SVGs feel less responsive than "lazy rendered" SVGs (because of the time it takes to request each individual SVG).

From this it seems to me like we could definitely switch to the "lazy rendering" approach, my reasons being:

1. If lazy loading is supported, this approach slightly underperforms on the initial load but it feels more response when scrolling.
2. If lazy loading is not supported, this approach is far better than actual lazy loading.


Any @simple-icons/maintainers have an opinion on this?

ericcornelissen committed:
Fix lazy loading icons in dark mode

ericcornelissen committed:
Implement lazy rendering with base64 encoded SVG in <img>

ericcornelissen committed:
External SVG with `filter: invert(1);`

ericcornelissen committed:
Revert to ba07d08dbba9aa3dc138c7681e55d09af808b003

ericcornelissen committed:
Remove console.log

ericcornelissen committed:
Un-async copy SVG callback

ericcornelissen committed:
Unskip copy SVG button test

ericcornelissen committed:
Merge branch 'master' into perf/img-loading-lazy

ericcornelissen committed:
Mock `atob` in copy.js test suite

ericcornelissen committed:
Merge branch 'master' into perf/img-loading-lazy

ericcornelissen committed:
Add missing Twitter logo to Share button in footer

ericcornelissen committed:
Add z-index: 1; to the grid-item__preview copy-button on-hover before
psuedo-element to ensure it is displayed in front of the preview SVG/
<img>. Without this it seems that, for some reason, in dark mode the
hover effect does not work as the ::before psuedo-element is rendered
behind the SVG.
